src/libopensc/card-starcos.c

```diff
@@ -1388,28 +1499,50 @@ static int starcos_set_security_env(sc_card_t *card,
 		}
 		ex_data->sec_ops        = SC_SEC_OPERATION_SIGN;
 		return SC_SUCCESS;
-	}
-
-	/* copy key reference, if present */
-	if (env->flags & SC_SEC_ENV_KEY_REF_PRESENT) {
-		if (env->flags & SC_SEC_ENV_KEY_REF_SYMMETRIC)
+	} else if(card->type == SC_CARD_TYPE_STARCOS_V3_2){
+		int temp = 0;
+		ex_data->fix_digestInfo = env->algorithm_flags; //TODO: set 0 if pkcs1 flag ?
+		switch (operation) {
+			case SC_SEC_OPERATION_DECIPHER:
+				sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x22, 0x41, 0xB8);
+				ex_data->sec_ops = SC_SEC_OPERATION_DECIPHER;
+				break;
+			case SC_SEC_OPERATION_SIGN:
+				/*TODO: for now: use internal authenticate to perform this operation
+				* (only works for RSA ?)
+				* not adequate setting found yet, no function to compare given,
+				* additional information missing */
+			case SC_SEC_OPERATION_AUTHENTICATE:
+				//see ref manual "internal authenticate"
+				sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x22, 0x41, 0xA4);
+				ex_data->sec_ops = SC_SEC_OPERATION_AUTHENTICATE;
+				break;
+			default:
+				return SC_ERROR_INVALID_ARGUMENTS;
+		}
+		//TODO not entirely correct but no appropriate flags ?
+		if (env->flags & SC_SEC_ENV_KEY_REF_SYMMETRIC){
+			//0x83: secret or public key for EXTERNAL AUTHENTICATE
+			//			or secret key for MUTUAL AUTHENTICATE
 			*p++ = 0x83;
-		else
+		}else{
+			//0x84 tag: private key for INTERNAL AUTHENTICATE
 			*p++ = 0x84;
+		}
+
 		*p++ = env->key_ref_len;
 		memcpy(p, env->key_ref, env->key_ref_len);
-		p += env->key_ref_len;
-	}
-	pp = p;
-	if (operation == SC_SEC_OPERATION_DECIPHER){
-		if (env->algorithm_flags & SC_ALGORITHM_RSA_PAD_PKCS1) {
-			*p++ = 0x80;
-			*p++ = 0x01;
-			*p++ = 0x02;
-		} else
-			return SC_ERROR_INVALID_ARGUMENTS;
-		sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x22, 0x81,
-		               0xb8);
+		// adapt last byte : FIXME DIRTY HACK (somehow works?)
+		p += env->key_ref_len -1;
```

Comment:
There is a discrepancy between provided key reference number by the card and reference number expected by it. This was tested on multiple cards and compared to other drivers which also receive the same numbers by the card. Even though e.g. a 0x84 is given as a key reference number when the card is read out, it produces an error when later specifying this exact key reference number for an operation. For an unknown reason, it always expects a number which is smaller by three, 0x81 in the example. The reason for this could not be found after thorough in depth search, so we left it at that. I see that there needs to be an additional check concerning a zero length key reference number, though.

Reply:
What is the exact key reference and the exact value you need to code in the MSE? What's the content of EF_KEYD?

I think understanding EF_KEYD is key in avoiding hacks like this. Please read the documentation and try to find some hints.

Reply:
I am fairly sure I tried the following, but I will look this up to confirm it:   
I was not able to access any EF_KEYD file as an additional source of information. Neither was this file accessed by any other driver, implementation or program which could be used as a reference. The documentation did not provide any further helpful information. Additionally, OpenSC chooses a key with the appropriate usage flags (see e.g. `pkcs15-framework.c:pkcs15_prkey_sign`, line 3773), whose reference number is written into the key-ref-pointer. This number is the exact same as the one which was initially read out from the Starcos Smart-Card. It is later not accepted as a valid ref-number anymore, the reported error code deeming the key not suitable or the number invalid. No information in the documentation suggests a change in or an adaption of the key reference number.

Reply:
Did you read the starcos reference manual? Did you search for *EF_KEYD*?

Sorry for being picky, but this hack is a little too problematic. Unfortunately, I currently don't have more time to implement this myself...

Reply:
I completely understand. I spent an extensive amount of time myself trying to find the reason for this. This also lead to the fact that I compared the complete communication process between driver and card for different implementations, none was making use of this EF_KEYD file or gaining any further relevant information. I am certain that the information found in the manual was not sufficient, but I am going to check that again tomorrow and come back to you.